copy/blob.go

```diff
@@ -128,8 +131,12 @@ func (ic *imageCopier) copyBlobFromStream(ctx context.Context, srcReader io.Read
 		}
 	}
 
-	if digestingReader.validationFailed { // Coverage: This should never happen.
-		return types.BlobInfo{}, fmt.Errorf("Internal error writing blob %s, digest verification failed but was ignored", srcInfo.Digest)
+	if digestingReader.validationFailed {
+		if !canRewriteDigest { // Coverage: This should never happen.
+			return types.BlobInfo{}, fmt.Errorf("Internal error writing blob %s, digest verification failed", srcInfo.Digest)
+		}
+		logrus.Warningf("Digest verification failed for blob %s but was ignored", srcInfo.Digest)
```

Comment:
That's a scary message. Can we explain why it's being ignored?

Reply:
the warning is outputted only when `canRewriteDigest == true`.  We haven't validated the digest earlier for partial images since we didn't pull the entire tarball, so for dedup purposes we use the TOC digest (which is not the DiffID).  When we copy the layer from the storage, we recompute the entire tarball so we can recalculate the correct DiffID and ignore the TOC Digest.

Reply:
If it expected it should be Info not warning.

Reply:
(Warning: At this point I don’t even have a strong opinion on the overall design. This is a drive-by comment.)

---

Like, why is this even validating a digest value when we know it is not going to match?

Please be precise in what the code does (or does not) track, or mean. `digestingReader` already provides `validationFailed`/`validationSucceeded`, and that can already be used to express “I have no opinion”. If something can’t be expressed with the existing data structures exactly, well, make that possible.

Please don’t create spaghetti where “`validationFailed` — oh but we didn’t _mean it_ here because another variable is set.”

---

(Actually if we don’t want to validate the blob’s digest… we shouldn’t even be computing it. But that’s a performance optimization I care very little about.)

Reply:
I've pushed a new version where I drop the warning.

I am open to suggestions on how to improve the implementation.  We still need to calculate the digest for the tarball because the one we are using with incremental images is the TOC digest (the only thing we can validate when we pull an image partially).

The idea is that when we generate the entire tarball, we can also calculate the diffID for the tarball, and use it instead of the digest we have in the store (the TOC digest).

How would we express "I have no opinion" with the current `digestReader`?  AFAICS, it will always return an error when the digest mismatches.  I don't think we can just ignore the error, since it could be related to something else.

Reply:
I was wrong, I don't really need to change the `digestReader` implementation and just skip the check when dealing with partial images.